### ISSUE TITLE
Cmake error for optional/

### ISSUE THREAD

**Issue by mycae** [libigl/libigl#703]
I am trying to build the compiled version of libigl, in order to ensure that I have a working build with my compiler and installed libraries (Debian stretch).

When doing so I get the following error (using Cmake 3.7.2):
```
CMake Error at CMakeLists.txt:52 (add_subdirectory):

    ${MY_HOMEDIR}/libigl/shared/cmake

  does not contain a CMakeLists.txt file.
```
this is due to the following line:
`
add_subdirectory("${PROJECT_SOURCE_DIR}/../shared/cmake" "libigl")
`

This is with checkout aa45ce.

Edit: 
I also get the same problem with shared/cmake/libigl.cmake, which references:
external/triangle
external/testgen/
external/glfw
external/CoMISo
external/embree

Thanks!

**Comment by jdumas** [comment]
Hi,

Please take a look at the [libigl-example-project](https://github.com/libigl/libigl-example-project) for a working example of how to build libigl in your own project.

**Comment by mycae** [comment]
Hmm,

Perhaps I should explain my situation a little better. That link appears to be a different repository, and not part of the main project repository (ie libigl/libigl).

I am considering incorporating this into my project, which is currently part of Fedora and Debian, but does not use IGL at this time. To integrate this, I need to be able to compile igl into a standalone installable module using the master git, make a .deb/.rpm, and then submit it to these distributions respective buildsystems. I can then apt-get/dnf install these at build time for my project, and thus use IGL.

I also need to be able to strip dependencies for which igl is not the canonical maintainer. I have done this for other libraries before.

I would like to be able to proceed with this, but to do so I think currently requires extensive changes to the buildsystem, so it builds "out-of-the-box", using the Cmake buildsystem upstream has selected

I'm happy to do some work here, however both the Fedora and Debian packaging guidelines have a strong "upstream first" approach, to try to make as small a changeset as possible to use the as-delivered upstream work (usually tarballs or tagged commits).

Any advice to getting the as-supplied build system to work would be greatly appreciated!

**Comment by jdumas** [comment]
Hi,

The current CMake build doesn't support what you mention at the moment. However, the core part of libigl is header-only, and only depends on Eigen, so it should be pretty easy to package it without going through CMake.

Right know we only provide a CMake script to build libigl as part of an existing project (e.g. via submodules), and the canonical example of that is the libigl-example-project that I have linked to. CMake is mostly a convenient way to compile each libigl module along with its dependencies, but strictly speaking it is not required for most modules.

If @alecjacobson and @danielepanozzo are interested in building .deb/.rpm packages for Debian/Fedora, and you are willing to help, I think it shouldn't be too hard to add a root `CMakeLists.txt` that builds libigl as a library (using CPack and whatnot). But I don't think it is a high priority for now.

**Comment by mycae** [comment]
Thanks for the very quick response.

Yes, I'm happy to assist with initial packaging (though I cannot provide time guarantees). I'm happy to wait on the other contributor's responses.

Thanks

**Comment by danielepanozzo** [comment]
Maybe I am missing something, but if your objective is to compile a shared library, this is very easy with the current cmake system. You can clone https://github.com/libigl/libigl-example-project and use this cmake:
```cmake
cmake_minimum_required(VERSION 3.1)
project(example)

set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${CMAKE_CURRENT_SOURCE_DIR}/cmake)

# libigl
option(LIBIGL_USE_STATIC_LIBRARY "Use libigl as static library" ON)
option(LIBIGL_WITH_ANTTWEAKBAR      "Use AntTweakBar"    OFF)
option(LIBIGL_WITH_CGAL             "Use CGAL"           OFF)
option(LIBIGL_WITH_COMISO           "Use CoMiso"         OFF)
option(LIBIGL_WITH_CORK             "Use Cork"           OFF)
option(LIBIGL_WITH_EMBREE           "Use Embree"         OFF)
option(LIBIGL_WITH_LIM              "Use LIM"            OFF)
option(LIBIGL_WITH_MATLAB           "Use Matlab"         OFF)
option(LIBIGL_WITH_MOSEK            "Use MOSEK"          OFF)
option(LIBIGL_WITH_OPENGL           "Use OpenGL"         ON)
option(LIBIGL_WITH_OPENGL_GLFW      "Use GLFW"           ON)
option(LIBIGL_WITH_PNG              "Use PNG"            OFF)
option(LIBIGL_WITH_PYTHON           "Use Python"         OFF)
option(LIBIGL_WITH_TETGEN           "Use Tetgen"         OFF)
option(LIBIGL_WITH_TRIANGLE         "Use Triangle"       OFF)
option(LIBIGL_WITH_VIEWER           "Use OpenGL viewer"  ON)
option(LIBIGL_WITH_XML              "Use XML"            OFF)

find_package(LIBIGL REQUIRED QUIET)

# Add your project files
file(GLOB SRCFILES *.cpp)
add_library(${PROJECT_NAME}_bin SHARED ${SRCFILES})

target_link_libraries(${PROJECT_NAME}_bin igl::core igl::opengl_glfw)
```

By turning on or off you can decide which dependencies you want included in the shared library.

**Comment by mycae** [comment]
Hi,

That doesn't work unless you first clone and install IGL. 
Here is my output : First clone example repository:

`$ git clone --depth=1 https://github.com/libigl/libigl-example-project.git`


Now cmake
```
$cmake .
<snip>
  libigl not found --- You can download it using:

  	git clone --recursive https://github.com/libigl/libigl.git
  /home/user/deb/libigl-example-project/../libigl (missing:
  LIBIGL_INCLUDE_DIR)
Call Stack (most recent call first):
  /usr/share/cmake-3.7/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindLIBIGL.cmake:30 (find_package_handle_standard_args)
  CMakeLists.txt:26 (find_package)


-- Configuring incomplete, errors occurred
```

This is not found as CMakeLists.txt is using find_package, which requires it to be in an installed location [1].  You have findIGL.cmake which looks up a few folders, and picks a few /usr/local/ locations, but still can't find my libigl (which is in a different BUILDROOT)

I first need to compile and install igl into /usr/include or /usr/local/include ... its a chicken and egg problem, as to install your example, I need to install igl, and I need the example to install the igl shared library via a .deb.  What you are describing is something where there are multiple interlocking tarballs from different repositories. 

This is unusual, but has been used by libraries such as harfbuzz and jogl/gluegen, and is generally quite tricky to work with. It is possible on Debian  [2], but usually its much easier to have a working CMake in the main repository or upstream tarball that runs some unit tests or compilation tests over the headers to ensure they work on the host.  The headers can be installed after those tests. For example, on my system I use Eigen3, instead of Eigen, so your `#include <Eigen/...>` doesn't work.

One also needs to ensure that these two different repos don't get out of sync, which I assume you are doing via CI or manually.

[1] https://cmake.org/cmake/help/v3.0/command/find_package.html
[2] https://raphaelhertzog.com/2010/09/07/how-to-use-multiple-upstream-tarballs-in-debian-source-packages/

**Comment by danielepanozzo** [comment]
If that helps, we could add the CMakeLists.txt I posted inside libigl/shared, so that you only need to clone libigl to run it. Or maybe I am still not understanding what you are trying to do.

**Comment by mycae** [comment]
Hi,

Thanks for persisting, and apologies for being some much trouble. 

To work here, ideally there is single repository with instructions for a working compile/install.

Unfortunately, copying your cmake file (verbatim) doesn't work as you define find_package, but not the required files for it in your primary repository: 
```
CMake Error at CMakeLists.txt:25 (find_package):
  By not providing "FindLIBIGL.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "LIBIGL", but
  CMake did not find one.

  Could not find a package configuration file provided by "LIBIGL" with any
  of the following names:

    LIBIGLConfig.cmake
    libigl-config.cmake

  Add the installation prefix of "LIBIGL" to CMAKE_PREFIX_PATH or set
  "LIBIGL_DIR" to a directory containing one of the above files.  If "LIBIGL"
  provides a separate development package or SDK, be sure it has been
  installed.
```

With time, I can debug this by hand, but it is much better if upstream does this, as it then becomes maintained, canonical, and in the format you want.

Does this make sense? Perhaps I am being a little unclear.

**Comment by mycae** [comment]
>  building the libigl sources along your project.
Unfortunately, this does not mesh with a typical building method for creating RPM/DEBs .

One installs a .deb that unpacks the shared library and headers. The user (on a their own personal machine, not on the build network) then creates their project to match against your library, i.e. by cloning your example project library.

The submitted deb/RPM must run some rudimentary compilation checks, such if a dependency shifts (as given by my eigen->eigen3 transition), this is detected at compilation stage, and a broken .deb is not submitted to the build network. 

One can imagine simply making a .deb that installs the headers only, and checks nothing at compilation stage. This would be very fragile, and not really a good plan.

Another option is to ship BOTH repositories at the same time, by picking a given commitID from each, making TWO .orig.tar.gz files on my machine (or equivalent in Fedora), and then making a multi-tarball commit (as per above. [2]).  This is not preferred, as it is quite unusual, and there is no guarantee that upstream has checked the same two commits I have picked.

The final option would be to ship TWO packages, but this is the circular dependency that I refer to. I cannot build each .deb/.rpm independently, as they require BOTH to be present, but cannot be, as they have different buildroots when being built by the build network.

**Comment by jdumas** [comment]
I'm sorry but if you goal is to distribute a compiled version of libigl as a .deb/.rpm package why don't you try to use CPack and exported targets to have CMake write the `LIBIGLConfig.cmake` that is needed to handle installed libraries?

Try to start with only the core library (export only the target igl::core), and get a working package from there on, and then we can work our way through the other modules. We've explained how to compile the libraries as part of any project, but why do you insist on shipping two packages, or that the libigl-example-project is interdependent with libigl?

To me it looks more like you there's something you don't get about how the CMake build works. You don't need a root CMakeLists.txt to use CPack and exported targets, which what you need to look at if you want to package/install libigl on your system.

**Comment by mycae** [comment]
I'm going to close this thread, as I think we are talking past one another, and we are not being productive.

Apologies again for taking up your time.

**Comment by jdumas** [comment]
Also, remember that libigl is also essentially a **header-only library**. This means that if you want to package it and install it on your system, it is enough to just copy the files to whatever folder you chose. CMake support is optional, and currently it is oriented for people who use libigl as a submodule in their project.

I don't even understand what you were trying to achieve here, as you haven't even stated your goal clearly.

**Comment by mycae** [comment]
Hi,

I think we are probably not helping one another, and that I am probably simply wasting your time.

I'd just like to emphasise, on behalf of your users, that It is not sufficient to copy the headers only, and that your statement is incorrect. This only works if the headers match the target system.

As a practical example of this, on my system, Eigen is not called Eigen, it is called Eigen3, as part of the Eigen2->3 transition. [1] See [2] for failure examples. It does  not work out-of-the box. I am unsure what your primary development platform is, which I think is the origin of why we cannot communicate effectively.

This is not only true for Eigen, but for any of your dependencies, which can change their API/ABI independently of your code.

[1] https://eigen.tuxfamily.org/dox/Eigen2ToEigen3.html
[2] https://stackoverflow.com/a/25537967

**Comment by jdumas** [comment]
Quoting the second link you posted:

> After exchanging a couple of comments it is clear that `/usr/include/eigen3/Eigen/Dense` should be include-able by `#include <Eigen/Dense>`, but not by `#include <eigen3/Eigen/Dense>`. Therefore, the addition of the command line option `-I /usr/include/eigen3` is mandatory.

I am a happy user of Archlinux, and I've been on Ubuntu for a while too, thank you very much. We only support the version of the libraries that are linked as submodule to libigl. Supporting different ABIs and different version of each dependency on different system is a huge amount of work (esp. if they change their API as you said). Given our limited resources, I do not think it is a reasonable thing to do at the moment.

**Comment by mycae** [comment]
Apologies if I am coming across as rude, this is not my intention. Thank you very kindly for your response.
